weaver/processes/esgf_process.py

```diff
@@ -0,0 +1,258 @@
+            dimensions.append(dimension)
+
+        if dimensions:
+            domain = cwt.Domain(
+                dimensions
+            )
+            return domain
+
+    def _check_required_inputs(self, workflow_inputs):
+        for required_input in self.required_inputs:
+            if required_input not in workflow_inputs:
+                raise ValueError("Missing required input: {}".format(required_input))
+
+    @staticmethod
+    def _get_files_urls(workflow_inputs):
+        # type: (JSON) -> List[Tuple[str, str]]
+        """Get all netcdf files from the cwl inputs"""
+        urls = []
+
+        files = workflow_inputs[InputNames.files]
+        if not isinstance(files, list):
+            files = [files]
+
+        for cwl_file in files:
+            if not cwl_file["class"] == "File":
+                raise ValueError("'{}' inputs must have a class named 'File'".format(InputNames.files))
+            location = cwl_file["location"]
+            if not location.startswith("http"):
+                raise ValueError("ESGF processes only support urls for files inputs.")
+            urls.append(location)
+        return urls
+
+    @staticmethod
+    def _get_variable(workflow_inputs):
+        # type: (JSON) -> str
+        """Get all netcdf files from the cwl inputs"""
+        if InputNames.variable not in workflow_inputs:
+            raise ValueError("Missing required input: variable")
+        return workflow_inputs[InputNames.variable]
+
+    def _run_process(self, api_key, inputs, domain=None):
+        # type: (str, List[cwt.Variable], Optional[cwt.Domain]) -> cwt.Process
+        """Run an ESGF process"""
+        LOGGER.debug("Connecting to ESGF WPS")
+
+        wps = cwt.WPSClient(self.provider, api_key=api_key, verify=True)
+        process = wps.processes(self.process)[0]
+
+        message = "Sending request."
+        LOGGER.debug(message)
+        self.update_status(message, Percent.SENDING, STATUS_RUNNING)
+
+        wps.execute(process, inputs=inputs, domain=domain)
+
+        LOGGER.debug("Waiting for result")
+
+        self._wait(process)
+
+        return process
+
+    def _wait(self, esgf_process, sleep_time=2):
+        # type: (cwt.Process, float) -> bool
+        """Wait for an ESGF process to finish, while reporting its status"""
+        status_history = set()
+
+        status_percent = [0]  # python 2 can't mutate nonlocal
+
+        def update_history():
+            status = esgf_process.status
+
+            if status not in status_history:
+                match = LAST_PERCENT_REGEX.match(status)
+                if match:
+                    status_percent[0] = int(match.group(1))
+                status_percent[0] = max(Percent.SENDING, status_percent[0])
+
+                status_history.add(status)
+
+                message = "ESGF status: " + status
+                LOGGER.debug(message)
+                self.update_status(message, status_percent[0], STATUS_RUNNING)
```

Comment:
that's great but there is no `self.percent`

Reply:
You're right, I mixed with wps_package. 
I would suggest only to replace by `status_percent = [0]` by `self.status_percent = 0` then an reuse it instead of the index hack.

Reply:
Do you mean this?

```
status_percent = 0  # python 2 can't mutate nonlocal

def update_history():
    status = esgf_process.status

    if status not in status_history:
        match = LAST_PERCENT_REGEX.match(status)
        if match:
            status_percent = int(match.group(1))
        status_percent = max(Percent.SENDING, status_percent)

        status_history.add(status)

        message = "ESGF status: " + status
        self.update_status(message, status_percent, STATUS_RUNNING)
```

Reply:
it should be a separate method that uses `def udpate_history(self)`
with the `self` context passed, you can use `self.status_percent` safely and the value will be update over various calls

Reply:
ok, would be done but not in this current PR